west: runners: blackmagicprobe: flash hex_file

tuna-f1sh opened the pull request:
The 'cfg.elf_file' is not signed, so `west flash` with a sysbuild will not run the application if MCUboot image signature checking is on. 

Using the 'cfg.hex_file' for `bmp_flash` resolves since as the signed hex is used when using sysbuild. Symbols are not required for flashing so the switching from elf to hex is not an issue.

Fixes #67803

tuna-f1sh committed:
west: runners: blackmagicprobe: flash hex_file

The cfg.elf_file is not signed, so `west flash` with a sysbuild will
not run the application if MCUboot image signature checking is on. Using
the cfg.hex_file for `bmp_flash` resolves since as the signed hex is
used when using sysbuild. Symbols are not required for flashing so the
switching from elf to hex is not an issue.

Signed-off-by: John Whittington <git@jbrengineering.co.uk>

tuna-f1sh commented:
Apologies for missing the test_blackmagicprobe.py tests. I'll push a fix.

tuna-f1sh commented:
Ok should be good to go now.

fabiobaltieri commented on scripts/west_commands/runners/blackmagicprobe.py:
```diff
@@ -150,16 +152,16 @@ def do_add_parser(cls, parser):
                             help='Assert SRST during connect? (default: no)')
 
     def bmp_flash(self, command, **kwargs):
-        if self.elf_file is None:
-            raise ValueError('Cannot debug; elf file is missing')
+        if self.hex_file is None:
+            raise ValueError('Cannot flash; hex file is missing')
```
Building the HEX file is optional (BUILD_OUTPUT_HEX), so at minimum this should still take the elf file if the hex is not specified, but also the output of gdb when flashing with the elf file is IMO nicer when using the elf than when using the hex file (it shows the section names rather than `Loading section .sec1...`.

How about:
- if hex file is not specified, use elf file
- if the hex and elf basename are the same, use the elf
- otherwise use the hex

that way the current behavior is unchanged unless there's a hint that the hex file has been further processed (by having a different name)

tuna-f1sh replied:
Thanks for review. I've added this fallback.

MeisterBob commented:
I had a PR on this a year ago, but then lost interest in BMP. Feel free to use the code from there if you want.
https://github.com/zephyrproject-rtos/zephyr/pull/51365

tuna-f1sh committed:
west: runners: blackmagicprobe: elf_file fallback

Building the HEX file is optional (CONFIG_BUILD_OUTPUT_HEX), so
`bmp_flash` will fallback to elf_file if missing. Additionally, to
maintain section names the HEX is only used if it is signed.

Signed-off-by: John Whittington <git@jbrengineering.co.uk>

tuna-f1sh commented:
@fabiobaltieri Yes, I've kept it simple and added a fallback to ELF if HEX is missing or it's not signed. I put it in a separate commit but can rebase if you'd prefer in one.

fabiobaltieri commented on scripts/west_commands/runners/blackmagicprobe.py:
```diff
@@ -150,16 +151,25 @@ def do_add_parser(cls, parser):
-        if self.elf_file is None:
-            raise ValueError('Cannot debug; elf file is missing')
+        # if hex file is not present or not signed, use the elf file
+        if self.hex_file is None:
```
could you swap the condition? I think it's easier to follow the logic if the case using `hex_file` is under `if self.hex_file:` rather than the nagated one

fabiobaltieri commented on scripts/west_commands/runners/blackmagicprobe.py:
```diff
@@ -116,6 +116,7 @@ def __init__(self, cfg, gdb_serial, connect_rst=False):
         #
         # https://github.com/zephyrproject-rtos/zephyr/issues/50789
         self.elf_file = Path(cfg.elf_file).as_posix()
+        self.hex_file = Path(cfg.hex_file).as_posix()
```
This needs an explicit check to see if `cfg.hex_file` is `None`, or it'll fail with `west build -t flash -- -DCONFIG_BUILD_OUTPUT_HEX=n`
